src/core/PathUtil.cpp

```diff
@@ -0,0 +1,148 @@
+			default                   : return QString("");
+		}
+		return QDir::cleanPath(loc) + "/";
+	}
+
+	QDir baseQDir (const Base base) { return QDir(baseLocation(base)); }
+
+	QString basePrefix(const Base base)
+	{
+		switch (base)
+		{
+			case Base::ProjectDir       : return QStringLiteral("userprojects:");
+			case Base::FactorySample    : return QStringLiteral("factorysample:");
+			case Base::UserSample       : return QStringLiteral("usersample:");
+			case Base::UserVST          : return QStringLiteral("uservst:");
+			case Base::Preset           : return QStringLiteral("preset:");
+			case Base::UserLADSPA       : return QStringLiteral("userladspa:");
+			case Base::DefaultLADSPA    : return QStringLiteral("defaultladspa:");
+			case Base::UserSoundfont    : return QStringLiteral("usersoundfont:");
+			case Base::DefaultSoundfont : return QStringLiteral("defaultsoundfont:");
+			case Base::UserGIG          : return QStringLiteral("usergig:");
+			case Base::DefaultGIG       : return QStringLiteral("defaultgig:");
+			default                   : return QStringLiteral("");
+		}
+	}
+
+	Base baseLookup(const QString & path)
+	{
+		for (auto base: relativeBases)
+		{
+			QString prefix = basePrefix(base);
+			if ( path.startsWith(prefix) ) { return base; }
+		}
+		return Base::Absolute;
+	}
+
+
+
+
+	QString stripPrefix(const QString & path)
+	{
+		return path.mid( basePrefix(baseLookup(path)).length() );
+	}
+
+	QString cleanName(const QString & path)
+	{
+		return stripPrefix(QFileInfo(path).baseName());
+	}
+
+
+
+
+	QString oldRelativeUpgrade(const QString & input)
+	{
+		if (input.isEmpty()) { return input; }
+
+		//Start by assuming that the file is a user sample
+		Base assumedBase = Base::UserSample;
+
+		//Check if it's a factory sample
+		QString factoryPath = baseLocation(Base::FactorySample) + input;
+		QFileInfo factoryInfo(factoryPath);
+		if (factoryInfo.exists()) { assumedBase = Base::FactorySample; }	
+
+		//Check if it's a VST
+		QString vstPath = baseLocation(Base::UserVST) + input;
+		QFileInfo vstInfo(vstPath);
+		if (vstInfo.exists()) { assumedBase = Base::UserVST; }
+
+		//Assume we've found the correct base location, return the full path
+		return basePrefix(assumedBase) + input;
+	}
+
+
+
+
+	QString toAbsolute(const QString & input)
+	{
+		//First, do no harm to absolute paths
+		QFileInfo inputFileInfo = QFileInfo(input);
+		if (inputFileInfo.isAbsolute()) { return input; }
+		//Next, handle old relative paths with no prefix
+		QString upgraded = input.contains(":") ? input : oldRelativeUpgrade(input);
+
+		Base base = baseLookup(upgraded);
+		return baseLocation(base) + upgraded.remove(0, basePrefix(base).length());
+	}
+
+	QString relativeOrAbsolute(const QString & input, const Base base)
+	{
+		if (input.isEmpty()) { return input; }
+		QString absolutePath = toAbsolute(input);
+		QString relativePath = baseQDir(base).relativeFilePath(absolutePath);
```

Comment:
If `base` is `Absolute`, `baseQDir(base)` will point to the working directory. It will result in undefined behaviors.

Reply:
Do you think this would fix it?
```diff
-QDir baseQDir (const Base base) { return QDir(baseLocation(base)); }
+QDir baseQDir (const Base base)
+{
+    if (base == Base::Absolute){ return QDir::root(); }
+    return QDir(baseLocation(base));
+}
```

Reply:
Better than before, but that may result in getting `foo/bar` for `/foo/bar` which will be interpreted as old-style relative path.